tap-version-14-specification.md

```diff
@@ -0,0 +1,741 @@
+  "severity": "fail",
+  "data": {
+    "got": {
+      "hostname": "peebles.example.com",
+      "address": null
+    },
+    "expected": {
+      "hostname": "peebles.example.com",
+      "address": "85.193.201.85"
+    }
+  }
+}
+```
+
+If a line occurs within the YAML document that is not indented, or if
+the TAP stream ends before the `/^ {2}\.\.\.$/` terminator line, or if the
+YAML document does not parse as YAML, then it MUST be treated as
+non-compliant data, and handled by the Harness the same as any other
+non-TAP lines.
+
+Currently (2015-10-06) the format of the data structure represented by
+a YAML block has not been standardized.  It is likely that a future
+version of TAP will standardize meanings for some of the fields in the
+YAML diagnostics.
+
+In order to adequately interpret and report the data within a test
+output's YAML block, a full YAML parser is required.  Specifying the
+YAML language is outside the scope of this document.  Harneses MAY
+ignore YAML blocks entirely if the inclusion of a full YAML parser
+would be overly difficult.
+
+## DIRECTIVES
+
+Directives are special notes that follow a # on the test line. Only
+two are currently defined: TODO and SKIP. Note that these two keywords
+are not case-sensitive.
+
+### TODO tests
+
+If the directive starts with `# TODO`, the test is counted as a todo
+test, and the text after TODO is the explanation.
+
+```
+not ok 13 # TODO bend space and time
+```
+
+Note that if the TODO has an explanation it must be separated from
+TODO by a space.  These tests represent a feature to be implemented or
+a bug to be fixed and act as something of an executable "things to do"
+list. They are not expected to succeed. Should a todo test point begin
+succeeding, the harness should report it as a bonus. This indicates
+that whatever you were supposed to do has been done and you should
+promote this to a normal test point.
+
+### Skipping tests
+
+If the directive starts with `# SKIP`, the test is counted as having
+been skipped. If the whole test file succeeds, the count of skipped
+tests is included in the generated output. The harness should report
+the text after # SKIP\S*\s+ as a reason for skipping.
+
+```
+ok 23 # skip Insufficient flogiston pressure.
+```
+
+Similarly, one can include an explanation in a Plan line, emitted if
+the test file is skipped completely:
+
+```
+1..0 # SKIP WWW::Mechanize not installed
+```
+
+## SUBTESTS
+
+A Subtest contains a block of test points grouped together.  This is
+useful for the output of test runners that collect several child
+processes' TAP streams, and output the result in TAP for another
+harness to interpret, or for tests with a predictable number of
+blocks, but where each block might not have a predictable number of
+tests.
+
+A Subtest is a TAP14 stream indented 4 spaces.  It has the following
+format:
+
+- A comment line with the name of the subtest: `/^ {4}# Subtest: (.*)$/`
```

Comment:
The reporters included in node-tap use this comment to introduce a suite of tests, which makes it much easier to mimic the output of mocha and rspec by getting the name of the suite up-front.  Without this, it's still easy enough to detect a subtest and distinguish from YAML (2 vs 4 space indentation, no `---` marker), but you have to wait until the end before reporting on the results in-progress next to the name.

I'm much more comfortable documenting what's already in practice.  It seems like the word "comment" is causing some valid concern here.  Would it help if we struck all the stuff about `#` lines being "comments", and instead just say that they're always "directives", and that unrecognized directives MUST be ignored?